alf/algorithms/mcts_algorithm.py

```diff
@@ -1013,21 +1020,31 @@ def _backup(self,
             discounts = self._discount**(path_lengths.unsqueeze(0) - steps)
             discounted_return = values.unsqueeze(0) * discounts
 
+        # [T, B] or [T, B, psims]
         valid = depth < path_lengths
         value_sum_delta = count * discounted_return[valid]
+        nodes = (B.expand_as(search_paths)[valid], search_paths[valid])
         if not parallel:
-            nodes = (B.expand(T, -1)[valid], search_paths[valid])
             trees.visit_count[nodes] += count
             trees.value_sum[nodes] += value_sum_delta
         else:
-            nodes = (B.expand(T, -1, psims)[valid], search_paths[valid])
-            nodes_i = nodes[0] * trees.visit_count.shape[1] + nodes[1]
-            # using scatter_add_ to handle possible duplicated nodes correctly
-            trees.visit_count.view(-1).scatter_add_(
-                0, nodes_i,
-                torch.tensor([count],
-                             dtype=trees.visit_count.dtype).expand_as(nodes_i))
-            trees.value_sum.view(-1).scatter_add_(0, nodes_i, value_sum_delta)
+            # nodes_i = nodes[0] * trees.visit_count.shape[1] + nodes[1]
+            # # using scatter_add_ to handle possible duplicated nodes correctly
+            # trees.visit_count.view(-1).scatter_add_(
+            #     0, nodes_i,
+            #     torch.tensor([count],
+            #                  dtype=trees.visit_count.dtype).expand_as(nodes_i))
+            # trees.value_sum.view(-1).scatter_add_(0, nodes_i, value_sum_delta)
+            # The scatter_add_ above is not deterministic. If deterministic behavior
+            # is needed, replace the above code with the following:
+            depth = depth.squeeze(-1)
```

Comment:
Both path are equivalent except that the first path is not deterministic, which means that there is still can be numerical difference due different order of arithmetic operations. In theory, the first path should be faster because the second path is a loop. But since this part of computation is small compare to other part of computation, so either way does not really affect speed much.